kmp-library/src/commonMain/kotlin/uk/co/lucasweb/aws/v4/signer/hash/ArrayExtensions.kt

```diff
@@ -0,0 +1,91 @@
+package uk.co.lucasweb.aws.v4.signer.hash
+
+// IntArray
+
+/**
+ * Converts the given int array into a byte array via big-endian conversion
+ * (1 int becomes 4 bytes).
+ * @return The converted array.
+ */
+internal fun IntArray.toByteArray(): ByteArray {
+    val array = ByteArray(this.size * 4)
+    for (i in this.indices) {
+        val bytes = this[i].toBytes()
+        array[i * 4] = bytes[0]
+        array[i * 4 + 1] = bytes[1]
+        array[i * 4 + 2] = bytes[2]
+        array[i * 4 + 3] = bytes[3]
+    }
+    return array
+}
+
+/**
+ * Copies [length] elements present inside [this] starting from [srcPos] into [dest] starting from [destPos].
+ * Thanks to manu0466
+ */
+internal fun IntArray.arrayCopy(srcPos: Int, dest: IntArray, destPos: Int, length: Int) {
+    // Make a deep copy avoiding errors when this == dest
+    val temp = this.copyOf()
+    (0 until length).forEach { dest[destPos + it] = temp[srcPos + it] }
```

Comment:
This idiom takes considerably longer than a `for` loop, depending on the size of the array this might cause a performance issue, reference: https://medium.com/@elye.project/kotlin-for-loop-vs-foreach-7eb594960333

```
intRange
println("ForLoop Time: " + measureNanoTime {
    for (i in 0 until 10000) { doSomething() }
})

println("ForEach Time: " + measureNanoTime {
    (0 until 10000).forEach { doSomething() }
})
Let’s looks at the result

ForLoop Time: 401199
ForEach Time: 5562190
``

Reply:
Since we're planning to add a dependency to KHash rather than copying the sources, I'd suggest to not spend time on changing that implementation.